core/src/main/java/org/apache/iceberg/LocationProviders.java

```diff
@@ -143,11 +143,11 @@ public String newDataLocation(PartitionSpec spec, StructLike partitionData, Stri
 
     @Override
     public String newDataLocation(String filename) {
-      int hash = HASH_FUNC.apply(filename);
+      String hash = computeHash(filename);
```

Comment:
I don't think that's a good idea.  If someone has a custom list operation, then they're relying on behaviors of the implementation and take the risk that something like this will break their implementation.

Reply:
what do you mean by custom list operation? Listing partitions based on the path after the random prefix?

Reply:
hi @jackye1995 if a object storage path has millions of files, S3 listing usually takes very long time since it has to go thru all the keys recursively.  This behavior can be improved by listing files with hash prefix in parallel like <location>/data/00, <location>/data/01, <location>/data/0a etc.

Reply:
+1 for what Daniel says.

Reply:
> if a object storage path has millions of files, S3 listing usually takes very long time since it has to go thru all the keys recursively. 

do you mean during orphan file removal? There could be better ways in making that work without listing if it's considered slow, like running a join against the inventory list.